sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BeamFnDataWriteRunner.java

```diff
@@ -72,87 +69,67 @@
           (FnDataReceiver) (FnDataReceiver<WindowedValue<InputT>>) runner::consume,
-          ((WindowedValueCoder<InputT>) runner.coder).getValueCoder());
+          ((WindowedValueCoder<InputT>) coder).getValueCoder());
 
-      context.addFinishBundleFunction(runner::close);
       return runner;
     }
   }
 
-  private final Endpoints.ApiServiceDescriptor apiServiceDescriptor;
-  private final String pTransformId;
-  private final Coder<WindowedValue<InputT>> coder;
-  private final BeamFnDataClient beamFnDataClientFactory;
-  private final Supplier<String> processBundleInstructionIdSupplier;
-
-  private CloseableFnDataReceiver<WindowedValue<InputT>> consumer;
-
-  BeamFnDataWriteRunner(
-      Supplier<Cache<?, ?>> cache,
-      String pTransformId,
-      RunnerApi.PTransform remoteWriteNode,
-      Supplier<String> processBundleInstructionIdSupplier,
-      Map<String, RunnerApi.Coder> coders,
-      BeamFnDataClient beamFnDataClientFactory,
-      BeamFnStateClient beamFnStateClient)
-      throws IOException {
-    this.pTransformId = pTransformId;
-    RemoteGrpcPort port = RemoteGrpcPortWrite.fromPTransform(remoteWriteNode).getPort();
-    this.apiServiceDescriptor = port.getApiServiceDescriptor();
-    this.beamFnDataClientFactory = beamFnDataClientFactory;
-    this.processBundleInstructionIdSupplier = processBundleInstructionIdSupplier;
-
-    RehydratedComponents components =
-        RehydratedComponents.forComponents(Components.newBuilder().putAllCoders(coders).build());
-    this.coder =
-        (Coder<WindowedValue<InputT>>)
-            CoderTranslation.fromProto(
-                coders.get(port.getCoderId()),
-                components,
-                new StateBackedIterableTranslationContext() {
-                  @Override
-                  public Supplier<Cache<?, ?>> getCache() {
-                    return cache;
-                  }
-
-                  @Override
-                  public BeamFnStateClient getStateClient() {
-                    return beamFnStateClient;
-                  }
-
-                  @Override
-                  public Supplier<String> getCurrentInstructionId() {
-                    return processBundleInstructionIdSupplier;
-                  }
-                });
-  }
+  private final BeamFnDataOutboundAggregator outboundAggregator;
+  private LogicalEndpoint endpoint;
 
-  public void registerForOutput() {
-    consumer =
-        beamFnDataClientFactory.send(
-            apiServiceDescriptor,
-            LogicalEndpoint.data(processBundleInstructionIdSupplier.get(), pTransformId),
-            coder);
+  BeamFnDataWriteRunner(BeamFnDataOutboundAggregator outboundAggregator) throws IOException {
+    this.outboundAggregator = outboundAggregator;
   }
 
-  public void close() throws Exception {
-    consumer.close();
+  public void registerForOutput(Supplier<LogicalEndpoint> outputLocation, Coder<?> coder) {
+    endpoint = outputLocation.get();
```

Comment:
If you provide the process bundle descriptor id supplier to the `BeamFnDataOutboundAggregator` you can register once during construction of the `BeamFnDataWriteRunner` with the `transform id` or `transform id and timer id` then the aggregator can own the logic of constructing the `LogicalEndpoint` if it still needs to.

Reply:
done.